
library(tidyverse)
library(ggrastr)
library(ggpubr)

# This data table is generated by fold_human_introns_observed_shuffled.py
intron_mfe_data <- read_tsv('CCDS_length_5000_introns_observed_shuffled_mfe_merged.txt.gz') %>%
  mutate(intron_length = nchar(intron_seq)) %>%
  mutate(observed_mfe_norm = observed_mfe/intron_length,
         shuffled_mfe_mean_norm = mean_shuffled_mfe/intron_length,
         shuffled_mfe_median_norm = median_shuffled_mfe/intron_length) %>%
  mutate(observed_shuffled_diff = observed_mfe_norm - shuffled_mfe_mean_norm) %>%
  filter(!is.na(shuffled_mfe_mean_norm)) %>%
  mutate(intron_id = paste(chrom, strand, intron_start+1, intron_end, sep = '_')) %>%
  select(intron_id, intron_length, observed_mfe_norm:observed_shuffled_diff)

# Plot comparison of normalized MFEs for 2 ICP0 introns versus distribution for human introns
icp0_intron1_mfe_norm <- -382.70/765
icp0_intron2_mfe_norm <- -50.00/136
ggplot(intron_mfe_data, aes(x=1, y=-observed_mfe_norm)) + geom_boxplot(outlier.shape = NA, fill='#00BBDB') +
  labs(x='Human introns', y='MFE (-kcal/mol/base)') +
  annotate(geom = 'point', x=1, y=-icp0_intron1_mfe_norm, size=1, color='red') +
  annotate(geom = 'point', x=1, y=-icp0_intron2_mfe_norm, size=1, color='red') +
  coord_cartesian(ylim=c(0, 0.61)) +
  theme_pubr() +
  theme(axis.text = element_text(size=7), axis.title = element_text(siz=7),
        axis.text.x = element_blank(), axis.ticks.x = element_blank())


# Total mapped read counts for normalization of each sample of an 8hr HSV1 infection timecourse
mapped_read_counts <- read_tsv('Rutkowski_HSV1_total_RNA_mapped_read_counts.txt') %>%
  group_by(source_rna, infection) %>%
  summarise(mapped_reads_total = sum(mapped_reads_total))

# Parse coverage of ICP0 and LAT exons/introns for each timecourse sample (derived using bedtools)
lat_icp0_read_counts <- tibble(infection = character(), replicate = character(), gene=character(), feature=character(), feature_type=character(), feature_num=character(), feature_length=numeric(), read_cov=numeric())
for (infection in c('NI', paste0(seq(2,8,2), 'hpi'))) {
  for (rep_num in paste0('rep', 1:2)){
      sample_lat_counts <- read_tsv(paste0('hsv1_exon_intron_read_counts/Total_', infection, '_', rep_num, '_HSV1_LAT_read_counts.txt'),
                                     col_names = c('chrom', 'start', 'end', 'info', 'score', 'strand', 'read_cov')) %>%
        separate(info, c('gene', 'feature'), sep = '_', remove = T) %>%
        mutate(feature_type = substr(feature, 1, nchar(feature)-1),
               feature_num = substr(feature, nchar(feature), nchar(feature)),
               feature_length = end-start) %>%
        mutate(infection=infection, replicate=rep_num)
      sample_icp0_counts <- read_tsv(paste0('hsv1_exon_intron_read_counts/Total_', infection, '_', rep_num, '_HSV1_ICP0_read_counts.txt'),
                                      col_names = c('chrom', 'start', 'end', 'info', 'score', 'strand', 'read_cov')) %>%
        separate(info, c('gene', 'feature'), sep = '_', remove = T) %>%
        mutate(feature_type = substr(feature, 1, nchar(feature)-1),
               feature_num = substr(feature, nchar(feature), nchar(feature)),
               feature_length = end-start) %>%
        mutate(infection=infection, replicate=rep_num)
      lat_icp0_read_counts <- bind_rows(lat_icp0_read_counts, sample_lat_counts, sample_icp0_counts)
  }
}

# Plot the exonic and intronic coverage for LAT and ICP0 over an 8hr infection timecourse
lat_icp0_read_counts_summ <- lat_icp0_read_counts %>%
  group_by(infection, gene, feature_type) %>%
  summarise(read_cov = sum(read_cov), feature_length = sum(feature_length)/2) %>% ungroup()
lat_icp0_read_counts_summ <- lat_icp0_read_counts_summ %>%
  left_join(mapped_read_counts, by='infection')
lat_icp0_read_counts_summ <- lat_icp0_read_counts_summ %>%
  mutate(reads_per_kb = read_cov/(feature_length/1000)) %>%
  mutate(read_cov_norm = reads_per_kb/(mapped_reads_total/1000000))
lat_icp0_read_counts_summ$infection <- lat_icp0_read_counts_summ$infection %>%
  recode('NI'=0, '2hpi'=2, '4hpi'=4, '6hpi'=6, '8hpi'=8)
lat_icp0_read_counts_summ <- lat_icp0_read_counts_summ %>%
  mutate(gene_feature = paste(gene, feature_type, sep = '_'))
ggplot(lat_icp0_read_counts_summ, aes(x=infection, y=read_cov_norm, color=gene_feature, fill=gene_feature)) +
  geom_point(size=0.5) + geom_line(linewidth=0.5) +
  labs(x='Hours post infection', y='Reads per kb per 10e6 mapped reads') +
  theme_pubr() +
  theme(legend.position = 'bottom', axis.text = element_text(size=7),
        axis.title = element_text(size=7), legend.text = element_text(size=7),
        legend.key.size = unit(3.5, 'mm'), legend.title = element_blank())
ggsave('results/hsv1_lat_icp0_exon_intron_total_cov_norm.pdf', units = 'mm', width = 45, height = 70)


# Plot lariat read level of ICP0 intron 1 over HSV1 infection timecourse
hsv1_lariats <- read_tsv('Rutkowski_HSV1_total_RNA_lariat_reads_merged.tsv')
hsv1_lariats_summ <- hsv1_lariats %>%
  group_by(hpi, rep) %>%
  summarise(lar_per_mil = n()/(max(total_mapped_reads)/1000000))
hsv1_lariats_summ <- hsv1_lariats_summ %>%
  bind_rows(data.frame(hpi=c('2hpi', '2hpi', '4hpi'), rep=c(1, 2, 1), lar_per_mil = c(0, 0 ,0)))
hsv1_lariats_summ_mean <- hsv1_lariats_summ %>%
  group_by(hpi) %>%
  summarise(lar_per_mil_mean = mean(lar_per_mil), lar_per_mil_se = sd(lar_per_mil)/sqrt(n()))
ggplot(hsv1_lariats_summ_mean, aes(x=hpi, y=lar_per_mil_mean, group=1)) + geom_point(size=0.5) + geom_line(linewidth=0.5) +
  geom_errorbar(aes(ymin=lar_per_mil_mean-lar_per_mil_se, ymax=lar_per_mil_mean+lar_per_mil_se), linewidth=0.5, width=0.3) +
  labs(x='Hours post infection', y='ICP0 intron 1 lariat reads\nper 10e6 mapped reads') +
  scale_x_discrete(breaks = paste0(c(2,4,6,8), 'hpi'), labels=c(2,4,6,8)) +
  theme_pubr() +
  theme(legend.position = 'none', axis.title = element_text(size=7),
        axis.text = element_text(size=7))
