
library(tidyverse)
library(ggpubr)
library(ggrastr)

# This data table is generated by alu_types_by_gene_region_all_genes.py
gene_region_alus <- read_tsv('hg19.gencode.basic.v45.gene_regions_with_alu_content.txt.gz')
gene_region_alus <- gene_region_alus %>%
  mutate(region_length = end-start+1,
         region_id = paste(chrom, strand, start, end, sep = '_')) %>%
  mutate(total_alu_per_kb = total_alu_count/(region_length/1000),
         ir_alu_per_kb = ir_alu_pair_count/(region_length/1000),
         has_alu = total_alu_count>0, has_ir_alu = ir_alu_pair_count>0)
gene_region_alus$gene_region <- factor(gene_region_alus$gene_region, levels = c('fivep_UTR', 'exon', 'intron', 'threep_UTR'))
gene_region_alus <- gene_region_alus %>%
  mutate(alu_category = ifelse(has_ir_alu, 'ir_alu', ifelse(has_alu & total_alu_count>1, 'non_ir_multi_alu', ifelse(total_alu_count==1, 'non_ir_single_alu', 'no_alu'))))
gene_region_alus$alu_category <- factor(gene_region_alus$alu_category, levels = c('no_alu', 'non_ir_single_alu', 'non_ir_multi_alu', 'ir_alu'))
alu_categories <- c('no_alu', 'non_ir_single_alu', 'non_ir_multi_alu', 'ir_alu')

# This data table is generated by fold_human_introns_observed_shuffled.py
intron_mfe_data <- read_tsv('CCDS_length_5000_introns_observed_shuffled_mfe_merged.txt.gz') %>%
  mutate(intron_length = nchar(intron_seq)) %>%
  mutate(observed_mfe_norm = observed_mfe/intron_length,
         shuffled_mfe_mean_norm = mean_shuffled_mfe/intron_length,
         shuffled_mfe_median_norm = median_shuffled_mfe/intron_length) %>%
  mutate(observed_shuffled_diff = observed_mfe_norm - shuffled_mfe_mean_norm) %>%
  filter(!is.na(shuffled_mfe_mean_norm)) %>%
  mutate(intron_id = paste(chrom, strand, intron_start+1, intron_end, sep = '_')) %>%
  select(intron_id, observed_mfe_norm:observed_shuffled_diff)
intron_alus_mfe <- gene_region_alus %>%
  filter(gene_region == 'intron') %>%
  mutate(intron_id = paste(chrom, strand, start, end, sep = '_')) %>%
  select(!transcript_id) %>% distinct() %>%
  right_join(intron_mfe_data, by='intron_id') 


# Plot MFE for observed and sequence-shuffled human introns
mfe_cor <- cor(intron_alus_mfe$observed_mfe_norm, intron_alus_mfe$shuffled_mfe_mean_norm)
mfe_r2 <- round(mfe_cor^2, 2)
ggplot(intron_alus_mfe, aes(x=observed_mfe_norm, y=shuffled_mfe_mean_norm)) +
  rasterize(geom_point(alpha=0.25, color='#00BCDB'), dpi = 600) +
  geom_abline(intercept = 0, slope = 1) +
  annotate(geom = 'text', label = bquote(r^2 == .(mfe_r2)), x=-0.8, y=-0.1, size=2) +
  labs(x='Observed MFE', y='Mean MFE of shuffled ensemble') +
  xlim(-1,0) + ylim(-1,0) +
  theme_pubr() +
  theme(axis.title = element_text(size=6), axis.text = element_text(size=5),
        aspect.ratio = 1)

# Plot MFE for observed and sequence-shuffled human introns subdivided by their Alu content
# (no Alu, 1 Alu, >1 Alu same orientation, >1 Alu IR orientation)
for (i in 1:4) {
  alu_cat <- alu_categories[[i]]
  intron_subset <- intron_alus_mfe %>%
    filter(alu_category == alu_cat)
  mfe_cor <- cor(intron_subset$observed_mfe_norm, intron_subset$shuffled_mfe_mean_norm)
  mfe_r2 <- round(mfe_cor^2, 2)
  ggplot(intron_subset, aes(x=observed_mfe_norm, y=shuffled_mfe_mean_norm)) +
    rasterize(geom_point(alpha=0.25, color='#00BCDB'), dpi = 600) +
    geom_abline(intercept = 0, slope = 1) +
    annotate(geom = 'text', label = bquote(r^2 == .(mfe_r2)), x=-0.8, y=-0.1, size=2) +
    labs(x='Observed MFE', y='Mean MFE of shuffled ensemble') +
    xlim(-1,0) + ylim(-1,0) +
    theme_pubr() +
    theme(axis.title = element_text(size=6), axis.text = element_text(size=5),
          aspect.ratio = 1)
}


# Plot the Alu content of each gene region (3' UTR, exon/CDS, intron and 5' UTR)
alu_prop_by_category <- gene_region_alus %>%
  select(!transcript_id) %>% distinct() %>%
  group_by(gene_region, alu_category) %>%
  summarise(region_count = n()) %>%
  drop_na(gene_region) %>%
  mutate(region_prop = region_count/sum(region_count)) %>%
  filter(alu_category != 'no_alu')
ggplot(alu_prop_by_category, aes(x=gene_region, y=100*region_prop, fill=alu_category)) +
  geom_col(position = 'stack') +
  labs(y='Gene regions with Alu (%)', fill='Alu category') +
  scale_x_discrete(breaks=c('exon', 'fivep_UTR', 'intron', 'threep_UTR'),
                   labels=c('Exons', '5\' UTR', 'Introns', '3\' UTR')) +
  scale_fill_viridis_d(breaks=c('non_ir_single_alu', 'non_ir_multi_alu', 'ir_alu'),
                       labels=c('1 Alu', '>1 Alu', 'IR Alu')) +
  theme_pubr() +
  theme(axis.title.x = element_blank(), axis.text.x = element_text(size=7, angle = 45, hjust = 0.9, vjust = 0.9),
        axis.title.y = element_text(size=7), axis.text.y = element_text(size=6),
        legend.position = 'bottom', legend.text = element_text(size=7),
        legend.title = element_text(size=7), legend.key.size = unit(4, 'mm'))


# Plot the change in lariat levels upon DBR1 knockout for introns that have IR Alu insertions
alu_intron_lariats <- read_tsv('lariat_levels_by_alu_category.txt')
first_lariat_mapped_read_counts <- read_tsv('C19_C22_293T_mapped_read_count.txt') %>%
  group_by(cell_line) %>% summarise(total_reads_exp = sum(mapped_reads)) %>% ungroup() %>% select(cell_line, total_reads_exp) %>% mutate(experiment = 1)
rescue_lariat_mapped_read_counts <- read_tsv('rescue_mapped_read_count.txt')%>%
  group_by(cell_line) %>% summarise(total_reads_exp = sum(mapped_reads)) %>% ungroup() %>% select(cell_line, total_reads_exp) %>% mutate(experiment = 2)
lariat_mapped_read_counts <- bind_rows(first_lariat_mapped_read_counts, rescue_lariat_mapped_read_counts) %>%
  pivot_wider(names_from = cell_line, values_from = total_reads_exp, names_prefix = 'total_reads_')
alu_intron_lariats <- alu_intron_lariats %>%
  left_join(lariat_mapped_read_counts, by='experiment')

alu_intron_lariat_summ <- alu_intron_lariats %>%
  group_by(experiment, alu_category) %>%
  summarise(lariat_reads_WT = sum(lariat_count_WT), lariat_reads_DBR1_KO = sum(lariat_count_DBR1_KO),
            total_reads_WT = max(total_reads_WT), total_reads_DBR1_KO = max(total_reads_DBR1_KO), lariat_count = n()) %>%
  ungroup() %>%
  mutate(lariat_norm_WT = lariat_reads_WT/(total_reads_WT/1000000),
         lariat_norm_DBR1_KO = lariat_reads_DBR1_KO/(total_reads_DBR1_KO/1000000)) %>%
  pivot_longer(c(lariat_norm_WT, lariat_norm_DBR1_KO), names_to = 'cell_line', names_prefix = 'lariat_norm_', values_to = 'lariat_norm') %>%
  group_by(alu_category, cell_line) %>%
  summarise(lariat_norm_mean = mean(lariat_norm), lariat_norm_se = sd(lariat_norm)/sqrt(n()))
alu_intron_lariat_summ$cell_line <- factor(alu_intron_lariat_summ$cell_line, levels = c('WT', 'DBR1_KO'))
ggplot(filter(alu_intron_lariat_summ, alu_category == 'ir_alu'), aes(x=cell_line, y=lariat_norm_mean)) + geom_col() +
  geom_errorbar(aes(ymin=lariat_norm_mean-lariat_norm_se, ymax=lariat_norm_mean+lariat_norm_se), width=0.3) +
  labs(y='IR Alu intron lariat reads\nper 10e6 mapped reads') +
  scale_x_discrete(breaks=c('DBR1_KO', 'WT'), labels=c('DBR1\nKO', 'WT')) +
  theme_pubr() +
  theme(legend.position = 'none', axis.title = element_text(size=7), axis.title.y = element_text(size=7),
        axis.text = element_text(size=7), axis.text.x = element_text(size=7),
        axis.title.x = element_blank(), plot.title = element_text(size=10, hjust=0.5), strip.text = element_text(size=8))

